Implement control mechanism for queue release

emanueleDiVizio opened the pull request:
## Motivation

While using this library, we encountered the need for halting the queue and resume it based on different conditions. In our app, we have two use cases for this. Since we see other people having the same issue (https://github.com/rgommezz/react-native-offline/issues/160), we thought to write a PR.
First, wait for the user to authenticate for the queue to be released. Second, we treat certain logic as blocking, so the queue will wait for the logic to be complete before resume. 
This is implemented with a simple semaphore mechanism.

## Test plan

The idea behind this PR is to have a semaphore mechanism for the queue. The logic is very similar to the one used to handle connectivity change. 
We have a semaphore variable `hasQueueBeenHalted`. 
To toggle the semaphore we've implemented the `queueSemaphoreChange(shouldHaltQueue: boolean)` action. If `shouldHaltQueue` is set to `true`, then `hasQueueBeenHalted` will be set to true and the queue will stop being released. 
If `shouldHaltQueue` is set to false, then `hasQueueBeenHalted` will be set to `false` and the queue will be resumed. This is all done only if either `isConnected` is true or it has been recently toggled.

I've added a couple of tests for the functionality.

Note, this is a working MVP at the moment. I'm pushing a PR already to gain early feedback and to let fellow developers benefit from this. Do please comment on this and share how it can be improved.

Closes #160

melvynhills commented:
Clean implementation 👌 

I personally would prefer naming this API in a simpler way, abstracting away the pattern that is used (semaphore), i.e. having actions called e.g. `enableQueue()` and `disableQueue()` or even `toggleQueue(bool)` instead of `queueSemaphoreChange(bool)`.

How would you compare this imperative "ON/OFF" API vs. something more declarative [like I did here](https://github.com/gomymove/react-native-offline/commit/fa1b7ffb03f2270b800f954948b2acfaedac28cb)? It allows me to just specify one single state selector that can enable/disable the queue automatically, as it is just a function of the current application state. 

Anyway, this looks like a nice & useful addition to the library!

emanueleDiVizio commented:
Hey, thanks for the feedback! I did think about the naming, more than happy to rename. How about `stopQueueRelease()` and `resumeQueueRelease()`?

Regarding the API, having a selector in the middleware config was indeed something we thought about, yet we didn't see this implementation. I think it's cleaner, thanks for the tip! 

Question, in your code I don't see any further check in `createReleaseQueue`, which as I understand would serve to halt the queue resume process if it's already running. Did you manage to find a way to do so without modifying `createReleaseQueue`?

emanueleDiVizio commented:
Alright, I've implemented it in a cleaner way following @melvynhills suggestion. I've opened a PR on our fork as I'm short on time to fix unit tests. If people want to take a look, here's our implementation with a deselector. https://github.com/evrythng/react-native-offline/pull/2/files. Thanks @melvynhills for the tip!

rgommezz commented on src/redux/actionCreators.js:
```diff
@@ -53,3 +54,14 @@ export const dismissActionsFromQueue = (
+  payload: shouldHaltQueue,
+});
+
+export default {
```
Nice! 💯 Exporting action creators is something the community asked for some time ago. Would you mind extending it to export all of them?

emanueleDiVizio replied:
Good point, I'll do it in another commit 👍

emanueleDiVizio replied:
Done 👍

melvynhills commented:
@rgommezz I think @emanueleDiVizio's plan is to do it in a more declarative way, as most people seem to like the idea of a state selector: https://github.com/evrythng/react-native-offline/pull/2/files

rgommezz commented:
Oh cool @melvynhills, sorry I missed the last comment and the separate fork.

I like that approach as well and would definitely be up to get this merged, could we make this happen @emanueleDiVizio? I can help with the tests if needed, just point your changes to this repo.

emanueleDiVizio commented:
So quick update, I've ended up using both mechanisms in our codebase, as I couldn't find a way to stop the queue with the selector if it's already been released. 
Code for both approaches being used at same time is here https://github.com/emanueleDiVizio/react-native-offline/commit/59244e6da01c961c2b2acf89d5d1177614fa4884.

I couldn't manage to update this PR yet, so code is in my fork for now. If you guys are happy to have both approaches available, I'll be glad to clean up code and update this PR. I'll also address @rgommezz comments :+1:

To speed this up, @rgommezz (and everyone else who's up for it) would you mind checking out https://github.com/emanueleDiVizio/react-native-offline/commit/59244e6da01c961c2b2acf89d5d1177614fa4884 to see if anything could be improved? I'll then open a PR here asap.

emanueleDiVizio commented on src/redux/actionTypes.js:
```diff
@@ -5,6 +5,7 @@ type ActionTypes = {|
   FETCH_OFFLINE_MODE: '@@network-connectivity/FETCH_OFFLINE_MODE',
   REMOVE_FROM_ACTION_QUEUE: '@@network-connectivity/REMOVE_FROM_ACTION_QUEUE',
   DISMISS_ACTIONS_FROM_QUEUE: '@@network-connectivity/DISMISS_ACTIONS_FROM_QUEUE',
+  QUEUE_SEMAPHORE_CHANGE: '@@network-connectivity/QUEUE_SEMAPHORE_CHANGE',
```
Done 👍

emanueleDiVizio commented on src/redux/actionCreators.js:
```diff
@@ -53,3 +54,14 @@ export const dismissActionsFromQueue = (
   payload: actionTrigger,
 });
+
+export const queueSemaphoreChange = (
```
Done 👍

emanueleDiVizio commented:
Okay great, I've managed to have both approaches side by side.

In the last two commits, I've implemented community suggestions and refactored a little bit. We also have tests for new functionality. Think I've ended up having both community suggestions and state selector into a single commit, apologies. Next thing I think would be to update the docs.

@rgommezz I've had a thought about moving the queue flag out of the reducer, yet I didn't manage to do it just yet. The point is that `createReleaseQueue` relies on redux state being updated in order to stop the queue, and I'm wondering if updating the variable in the parent closure would be seen by `createReleaseQueue` straight away as well. Probably yes, yet I'm prioritising not breaking current functionality over refactor, as we're using this in our app.

If anyone else (or me later on) manages to move out the `isQueuePaused` flag from the reducer, I'll be more than happy.

Let me know if there's any other way I can help!

emanueleDiVizio commented:
Alright, I've fixed it, tests are now passing. Semaphore is working and deselector is working. Only thing is if queue deselctor changes from true to false, the queue is not automatically released. Apart from this, doc update is last thing left I reckon. @rgommezz If everything looks good, I can provide docs :+1:

emanueleDiVizio commented:
Btw @rgommezz, I can't fathom why but I have been using a DAZN pen for the past few days :') I've found it in my office...

MakhouT commented:
I am happy to test it if you can give me some pointers on how to use it. I am using redux-saga

emanueleDiVizio commented:
So we have two ways to control the queue. One is with ` shouldDequeueSelector` and the other with semaphore actions. `shouldDequeueSelector` is a `func` returning `bool`. If `true`, then the queue can be released, if `false` queue release is prevented. Note, this only works *before* the queue release is initiated. If `shouldDequeueSelector`suddenly returns `false` while the queue is getting released then it won't be stopped. If we want to halt the queue *while* it's being released we have to dispatch a`CHANGE_QUEUE_SEMAPHORE` action, with 'RED' payload.
This action works in the following way: if semaphore is `RED`, queue release is halted, if `GREEN` then it can go. Let me know if more info is needed.

MakhouT commented:
I'll try it out this week if I get some time. When I've tested it, I'll get back to you!

emanueleDiVizio commented:
Sweet, thank you @MakhouT! @rgommezz is there anything else to do to make this happen?

rgommezz commented:
@emanueleDiVizio I briefly glanced at it during the weekend and it look quite solid, well done! 🎉 I'll fully re-review tomorrow

rgommezz commented on src/redux/createNetworkMiddleware.js:
```diff
@@ -70,11 +71,19 @@ function didComeBackOnline(action, wasConnected) {
+function didQueueResume(action, isQueuePaused) {
+  return (
+    action.type === networkActionTypes.CHANGE_QUEUE_SEMAPHORE &&
+    !isQueuePaused &&
```
Wouldn't this entail that the queue was paused before?

So the condition would be:

```js
  return (
    action.type === networkActionTypes.CHANGE_QUEUE_SEMAPHORE &&
    isQueuePaused && // shouldn't it be like this?
    action.payload === 'GREEN'
  );
```

emanueleDiVizio replied:
I believe we're checking that the queue was paused *before* this action is received, since by the time the middleware receives `CHANGE_QUEUE_SEMAPHORE`, `isQueuePaused` hasn't been updated yet. So I think that `!isQueuePaused` is the correct check. Does that sound right?

rgommezz replied:
Right, that's exactly my point.

As you are saying, we're checking that the queue was paused before this action is received, so `isQueuePaused === true` at the moment the action arrives to the middleware. The semaphore action indicates that we should `unblock/unpause` the queue.

So for me, `didQueueResume` means that:
- And action of type `action.type === networkActionTypes.CHANGE_QUEUE_SEMAPHORE` is about to be dispatched.
- The queue is currently paused: `isQueuePaused === true`
- The payload is `SemaphoreColor.GREEN`, meaning we should unpause/unblock the queue

emanueleDiVizio replied:
Done 👍

rgommezz replied:
I still don't see the fix @emanueleDiVizio 🤔

emanueleDiVizio replied:
True, I've thought about it. Not sure how it plays with Flow tho, since we've defined enums for it... Should I just add a SemaphoreColors object without flow?

emanueleDiVizio replied:
Can confirm fix is in d0393a6

emanueleDiVizio replied:
Alright done, do we also want to export `SEMAPHORE_COLOR` const?

rgommezz replied:
You changed a different function on https://github.com/rgommezz/react-native-offline/commit/d0393a648443970849b1b8d825eee1f6b1c9cd72 😕

rgommezz replied:
> True, I've thought about it. Not sure how it plays with Flow tho, since we've defined enums for it... Should I just add a SemaphoreColors object without flow?

Yes, I think that's better. Once we migrate to Typescript in the future, we'll be able to leverage Enums

emanueleDiVizio replied:
Uh, sorry about that! Will fix this today!

rgommezz commented on src/redux/createNetworkMiddleware.js:
```diff
@@ -114,7 +124,14 @@ function createNetworkMiddleware({
-    if (isBackOnline) {
+    const hasQueueBeenResumed = didQueueResume(action, isQueuePaused);
+
+    const shouldDequeue =
```
What about this regrouping, since `isBackOnline` by itself was already a valid condition?

```js
    const shouldDequeue =
      (isBackOnline || (isConnected && hasQueueBeenResumed)) &&
      shouldDequeueSelector(getState());
```

emanueleDiVizio replied:
Good idea, I'll implement this :+1:

rgommezz commented on src/redux/createNetworkMiddleware.js:
```diff
@@ -89,11 +98,12 @@ function createNetworkMiddleware({
     next: (action: any) => void,
   ) => (action: any) => {
-    const { isConnected, actionQueue } = getState().network;
+    const { isConnected, actionQueue, isQueuePaused } = getState().network;
```
Tiny opt. Let's keep the state in a variable right away since we are calling `getState()` twice.

emanueleDiVizio replied:
So the reason for calling getState inside the `releaseQueue` loop is that this way the queue release function always has the latest `state` on every iteration. This way, when `CHANGE_QUEUE_SEMAPHORE` and  `CONNECTION_CHANGE` are dispatched the new state is reflected straight away inside `releaseQueue`, since at every iteration we call `getState()`. On the other hand, if we pass down `isQueuePaused` and `isConnected` from middleware function their values will only reflect the state when the `releaseQueue` was created, and will not update on further state change while the queue is being released. Does that sound right?

rgommezz commented on src/types.js:
```diff
@@ -4,6 +4,8 @@ export type State = {
   isConnected: boolean,
 };
 
+export type SemaphoreColor = 'RED' | 'GREEN';
```
💯

emanueleDiVizio commented:
Done, I've removed the `yarn.lock` and regrouped the `shouldDeque` expression. I've left a couple of comments :+1:

rgommezz commented:
Btw @emanueleDiVizio, I can't merge this without having 100% coverage. You can take a look by running `npm run test:coverage` locally. It seems that function (`didQueueResume`) I spotted that seemed faulty is not properly tested and that's why the tests didn't catch that.

emanueleDiVizio commented:
Alright good news, I've fixed `didComeBackOnline` and `didQueueResume` functions and I've brought back coverage to 100%! I believe we're close, let me know if there's anything else I can do 👍

rgommezz commented on test/reducer.test.js:
```diff
@@ -383,3 +406,11 @@ describe('networkSelector', () => {
   });
 });
+
+describe('network reducer config', () => {
```
What's the purpose of this test?

emanueleDiVizio replied:
I think this is a leftover from previous code, I'll remove it. Good catch :+1:

emanueleDiVizio commented on test/createNetworkMiddleware.test.js:
```diff
@@ -396,12 +418,35 @@ describe('createReleaseQueue', () => {
+    expect(mockDispatch).toHaveBeenNthCalledWith(4, 'bar');
+  });
+
+  it('does not empty the queue if we are online and dequeue selector returns false', async () => {
```
Yeah that's the correct implementation, I'll address this :+1:

emanueleDiVizio commented on src/redux/createReducer.js:
```diff
@@ -8,11 +8,13 @@ import type {
   NetworkState,
+  SemaphoreColor,
 } from '../types';
 
```
Have to keep `SemaphoreColor` for typing in `handleChangeQueueSemaphore` (line 88). I'll import `SEMAPHORE_COLOR` as well.

emanueleDiVizio commented:
I've addressed comments :+1: 
On a side note, I believe I've found a way to stop the queue with the deselector even while it's being released. See code for `releaseQueue` on df12a54 for more detail. 
Only thing, I wasn't able to test this behaviour as I couldn't get the state to change while the queue is being released. I've tried using Promise.all([]) but didn't have any success. It's worth noting that there's no test for the queue being stopped while being released even with the queue semaphore, as I've encountered the same trobule.
@rgommezz could you look into df12a54? Maybe we should think of a way to test the queue being paused midway...

emanueleDiVizio commented:
Yuppie! Thanks for the great support @rgommezz! 
On a final note, I believe that dequeueSelector now works even while the queue is being released, but I haven't been able to test it in a real scenario as we moved away from this implementation in our codebase. 
@MakhouT since I believe you're planning to test it, would you mind testing that deselector and semaphore work while the queue is being released? We might want to update the docs if that's the case :+1:

rgommezz commented:
@all-contributors please add @emanueleDiVizio for code and doc

MakhouT commented:
Nice work guys! Unfortunately I wasn't able to test it yet, I'll try to get to it ASAP, as we also need this in our app. Will provide feedback when it got tested from my side.
